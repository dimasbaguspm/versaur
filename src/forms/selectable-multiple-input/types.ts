import type { ReactNode, InputHTMLAttributes } from 'react'

/**
 * Checkbox placement position
 */
type CheckboxPlacement = 'top' | 'center' | 'bottom'

/**
 * Props for SelectableMultipleInput
 */
export interface SelectableMultipleInputProps
  extends Omit<InputHTMLAttributes<HTMLInputElement>, 'type' | 'children'> {
  /**
   * The value of the checkbox input
   */
  value: string
  /**
   * The content to display next to the checkbox input
   * Can be a ReactNode or a function that receives the checked state
   */
  children: ReactNode | ((checked: boolean) => ReactNode)
  /**
   * Whether the input is checked (controlled)
   */
  checked: boolean
  /**
   * The placement of the checkbox relative to the content
   * @default 'center'
   */
  checkboxPlacement?: CheckboxPlacement
  /**
   * Whether to hide the checkbox
   * When true, only the content is displayed and the checked state is managed externally
   * @default false
   */
  hideCheckbox?: boolean
}
