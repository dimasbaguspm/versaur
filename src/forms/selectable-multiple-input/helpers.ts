import { cva } from 'class-variance-authority'

export const selectableMultipleInputRootClass = cva(
  'flex items-center gap-3 cursor-pointer border-b border-border select-none p-4 transition-colors hover:bg-neutral-light',
  {
    variants: {
      checked: {
        true: '',
        false: '',
      },
      disabled: {
        true: 'opacity-50 cursor-not-allowed',
        false: '',
      },
    },
    defaultVariants: {
      checked: false,
      disabled: false,
    },
  }
)

/**
 * Mobile-friendly box and focus/active/checked state classes for the checkbox box
 */
export const selectableBoxClass = cva(
  'flex h-6 w-6 items-center justify-center rounded-sm overflow-hidden border border-border transition-colors duration-150 focus-visible:ring-2 focus-visible:ring-primary focus-visible:ring-offset-2',
  {
    variants: {
      checked: {
        true: 'bg-primary',
        false: 'border-border bg-white',
      },
      disabled: {
        true: 'opacity-50',
        false: 'active:scale-95',
      },
    },
    defaultVariants: {
      checked: false,
      disabled: false,
    },
  }
)
