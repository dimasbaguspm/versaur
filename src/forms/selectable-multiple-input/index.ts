export { SelectableMultipleInput } from './selectable-multiple-input'
export type * from './types'
