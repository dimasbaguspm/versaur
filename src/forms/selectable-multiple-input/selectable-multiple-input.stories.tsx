import type { Meta, StoryObj } from '@storybook/react'
import { useState } from 'react'
import { SelectableMultipleInput } from './selectable-multiple-input'

/**
 * SelectableMultipleInput allows users to select multiple options using checkboxes with custom content.
 * It supports both static content and dynamic rendering through function-as-children pattern.
 */
const meta: Meta<typeof SelectableMultipleInput> = {
  title: 'Forms/SelectableMultipleInput',
  component: SelectableMultipleInput,
}
export default meta

type Story = StoryObj<typeof SelectableMultipleInput>

/**
 * Basic usage with simple text content
 */
export const Basic: Story = {
  render: function Basic() {
    const [selected, setSelected] = useState<string[]>([])
    const options = [
      { value: 'apple', label: 'Apple' },
      { value: 'banana', label: 'Banana' },
      { value: 'orange', label: 'Orange' },
    ]
    return (
      <div className='flex flex-col'>
        {options.map(opt => (
          <SelectableMultipleInput
            key={opt.value}
            value={opt.value}
            checked={selected.includes(opt.value)}
            onChange={e => {
              setSelected(sel =>
                e.target.checked
                  ? [...sel, opt.value]
                  : sel.filter(v => v !== opt.value)
              )
            }}
          >
            {opt.label}
          </SelectableMultipleInput>
        ))}
      </div>
    )
  },
}

/**
 * Dynamic content using function-as-children pattern.
 * The content changes based on the checked state.
 */
export const DynamicContent: Story = {
  render: function DynamicContent() {
    const [selected, setSelected] = useState<string[]>(['premium'])
    const plans = [
      { value: 'free', name: 'Free Plan', price: '$0/month' },
      { value: 'premium', name: 'Premium Plan', price: '$9.99/month' },
      { value: 'enterprise', name: 'Enterprise Plan', price: '$29.99/month' },
    ]
    return (
      <div className='flex flex-col'>
        {plans.map(plan => (
          <SelectableMultipleInput
            key={plan.value}
            value={plan.value}
            checked={selected.includes(plan.value)}
            onChange={e => {
              setSelected(sel =>
                e.target.checked
                  ? [...sel, plan.value]
                  : sel.filter(v => v !== plan.value)
              )
            }}
          >
            {checked => (
              <div>
                <p className='font-medium'>{plan.name}</p>
                <p className='text-sm text-foreground-light'>
                  {checked ? `✓ ${plan.price} - Selected` : plan.price}
                </p>
              </div>
            )}
          </SelectableMultipleInput>
        ))}
      </div>
    )
  },
}

/**
 * Custom checkbox placement for multi-line content.
 * Demonstrates top, center (default), and bottom alignments.
 */
export const CheckboxPlacement: Story = {
  render: function CheckboxPlacement() {
    const [selected, setSelected] = useState<string[]>([])
    return (
      <div className='flex flex-col'>
        <SelectableMultipleInput
          value='top'
          checked={selected.includes('top')}
          onChange={e => {
            setSelected(sel =>
              e.target.checked ? [...sel, 'top'] : sel.filter(v => v !== 'top')
            )
          }}
          checkboxPlacement='top'
        >
          <div>
            <p className='font-medium'>Top Aligned</p>
            <p className='text-sm text-foreground-light'>
              The checkbox is aligned to the top of this multi-line content.
            </p>
            <p className='text-sm text-foreground-light'>
              This is useful when you have tall content.
            </p>
          </div>
        </SelectableMultipleInput>
        <SelectableMultipleInput
          value='center'
          checked={selected.includes('center')}
          onChange={e => {
            setSelected(sel =>
              e.target.checked
                ? [...sel, 'center']
                : sel.filter(v => v !== 'center')
            )
          }}
          checkboxPlacement='center'
        >
          <div>
            <p className='font-medium'>Center Aligned (Default)</p>
            <p className='text-sm text-foreground-light'>
              The checkbox is vertically centered with the content.
            </p>
          </div>
        </SelectableMultipleInput>
        <SelectableMultipleInput
          value='bottom'
          checked={selected.includes('bottom')}
          onChange={e => {
            setSelected(sel =>
              e.target.checked
                ? [...sel, 'bottom']
                : sel.filter(v => v !== 'bottom')
            )
          }}
          checkboxPlacement='bottom'
        >
          <div>
            <p className='font-medium'>Bottom Aligned</p>
            <p className='text-sm text-foreground-light'>
              The checkbox is aligned to the bottom of this multi-line content.
            </p>
            <p className='text-sm text-foreground-light'>
              Great for content with varying heights.
            </p>
          </div>
        </SelectableMultipleInput>
      </div>
    )
  },
}

/**
 * Hide the checkbox indicator for fully custom rendering.
 * Useful when you want complete control over the visual presentation.
 */
export const HiddenCheckbox: Story = {
  render: function HiddenCheckbox() {
    const [selected, setSelected] = useState<string[]>(['option2'])
    const options = [
      { value: 'option1', label: 'Option 1' },
      { value: 'option2', label: 'Option 2' },
      { value: 'option3', label: 'Option 3' },
    ]
    return (
      <div className='flex flex-col'>
        {options.map(opt => (
          <SelectableMultipleInput
            key={opt.value}
            value={opt.value}
            checked={selected.includes(opt.value)}
            onChange={e => {
              setSelected(sel =>
                e.target.checked
                  ? [...sel, opt.value]
                  : sel.filter(v => v !== opt.value)
              )
            }}
            hideCheckbox
          >
            {checked => (
              <div
                className={`w-full p-3 rounded transition-all ${
                  checked
                    ? 'bg-primary-soft border-2 border-primary'
                    : 'bg-neutral-light border-2 border-transparent'
                }`}
              >
                <p className='font-medium'>
                  {checked ? '✓ ' : ''}
                  {opt.label}
                </p>
                <p className='text-sm text-foreground-light'>
                  {checked ? 'Selected' : 'Click to select'}
                </p>
              </div>
            )}
          </SelectableMultipleInput>
        ))}
      </div>
    )
  },
}

/**
 * Disabled state demonstration
 */
export const Disabled: Story = {
  render: function Disabled() {
    return (
      <div className='flex flex-col'>
        <SelectableMultipleInput
          value='disabled-unchecked'
          checked={false}
          disabled
        >
          Disabled Unchecked
        </SelectableMultipleInput>
        <SelectableMultipleInput
          value='disabled-checked'
          checked={true}
          disabled
        >
          Disabled Checked
        </SelectableMultipleInput>
      </div>
    )
  },
}
