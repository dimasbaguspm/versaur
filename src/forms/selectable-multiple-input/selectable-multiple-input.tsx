import { forwardRef, useId } from 'react'
import type { SelectableMultipleInputProps } from './types'
import { cn } from '@/utils/cn'
import { Icon } from '@/primitive/icon'
import { Check } from 'lucide-react'
import { selectableBoxClass, selectableMultipleInputRootClass } from './helpers'

/**
 * SelectableMultipleInput is a checkbox input with a custom label and checked icon
 * @see https://www.w3.org/WAI/ARIA/apg/patterns/checkbox/
 */
export const SelectableMultipleInput = forwardRef<
  HTMLInputElement,
  SelectableMultipleInputProps
>(({ value, label, checked, id, className, disabled, ...props }, ref) => {
  const computedId = useId()
  const inputId = id || computedId

  return (
    <label
      htmlFor={inputId}
      className={cn(
        selectableMultipleInputRootClass({
          checked: !!checked,
          disabled: !!disabled,
        }),
        className
      )}
      aria-checked={checked}
      role='checkbox'
      aria-label={`Select ${value}`}
      tabIndex={disabled ? -1 : 0}
    >
      <input
        ref={ref}
        id={inputId}
        type='checkbox'
        value={value}
        checked={checked}
        disabled={disabled}
        className='sr-only'
        aria-hidden
        aria-checked={checked}
        {...props}
      />
      <span
        className={selectableBoxClass({
          checked: !!checked,
          disabled: !!disabled,
        })}
        aria-hidden='true'
      >
        {checked ? <Icon as={Check} color='neutral' size='md' /> : null}
      </span>
      <span className='flex-1 min-w-0 text-base text-foreground'>{label}</span>
    </label>
  )
})
