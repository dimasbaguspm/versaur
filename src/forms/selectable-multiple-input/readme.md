# SelectableMultipleInput

A multiple-choice input component for Versaur, following the same UI and API as SelectableSingleInput but supporting multiple selections (checkbox).

## Usage

```tsx
import { SelectableMultipleInput } from '@/forms/selectable-multiple-input'

<SelectableMultipleInput
  value="option1"
  label="Option 1"
  checked={selected.includes('option1')}
  onChange={handleChange}
/>
```

- Uses the same style and accessibility patterns as SelectableSingleInput
- Use for multiple-choice selection UIs
