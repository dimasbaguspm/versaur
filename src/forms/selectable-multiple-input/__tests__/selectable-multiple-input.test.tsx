import { render, screen, fireEvent } from '@testing-library/react'
import { composeStories } from '@storybook/react'
import * as stories from '../selectable-multiple-input.stories'

describe('SelectableMultipleInput', () => {
  const { Basic, DynamicContent, CheckboxPlacement, HiddenCheckbox, Disabled } =
    composeStories(stories)

  it('renders basic options and allows multiple selections', () => {
    const { getByLabelText, asFragment } = render(<Basic />)
    // Snapshot for html output
    expect(asFragment()).toMatchSnapshot()

    // All options should be rendered
    expect(getByLabelText('Select apple')).toBeInTheDocument()
    expect(getByLabelText('Select banana')).toBeInTheDocument()
    expect(getByLabelText('Select orange')).toBeInTheDocument()

    // Nothing should be checked by default
    const apple = getByLabelText('Select apple') as HTMLLabelElement
    expect(apple.getAttribute('aria-checked')).toBe('false')

    // Simulate selecting apple
    fireEvent.click(apple)
    // After click, apple should be checked
    expect(apple.getAttribute('aria-checked')).toBe('true')

    // Simulate selecting banana (multiple selection)
    const banana = getByLabelText('Select banana') as HTMLLabelElement
    fireEvent.click(banana)
    // Both should now be checked
    expect(apple.getAttribute('aria-checked')).toBe('true')
    expect(banana.getAttribute('aria-checked')).toBe('true')
  })

  it('renders dynamic content based on checked state', () => {
    const { getByLabelText } = render(<DynamicContent />)

    // Premium should be checked by default
    const premium = getByLabelText('Select premium') as HTMLLabelElement
    expect(premium.getAttribute('aria-checked')).toBe('true')
    expect(premium.textContent).toContain('✓ $9.99/month - Selected')

    // Free should not be checked
    const free = getByLabelText('Select free') as HTMLLabelElement
    expect(free.getAttribute('aria-checked')).toBe('false')
    expect(free.textContent).toContain('$0/month')
    expect(free.textContent).not.toContain('✓')

    // Click free option (multiple selection)
    fireEvent.click(free)
    expect(free.getAttribute('aria-checked')).toBe('true')
    expect(free.textContent).toContain('✓ $0/month - Selected')
    // Premium should still be checked
    expect(premium.getAttribute('aria-checked')).toBe('true')
  })

  it('renders checkbox at different placements', () => {
    const { getByLabelText } = render(<CheckboxPlacement />)

    const topOption = getByLabelText('Select top') as HTMLLabelElement
    const centerOption = getByLabelText('Select center') as HTMLLabelElement
    const bottomOption = getByLabelText('Select bottom') as HTMLLabelElement

    // Check that all options are rendered
    expect(topOption).toBeInTheDocument()
    expect(centerOption).toBeInTheDocument()
    expect(bottomOption).toBeInTheDocument()

    // Nothing should be checked by default
    expect(topOption.getAttribute('aria-checked')).toBe('false')
    expect(centerOption.getAttribute('aria-checked')).toBe('false')
    expect(bottomOption.getAttribute('aria-checked')).toBe('false')

    // Verify placement classes
    expect(topOption).toHaveClass('items-start')
    expect(centerOption).toHaveClass('items-center')
    expect(bottomOption).toHaveClass('items-end')
  })

  it('renders without checkbox when hideCheckbox is true', () => {
    const { container } = render(<HiddenCheckbox />)

    // Should not find checkbox indicators (the span with the checkbox styling)
    const checkboxIndicators = container.querySelectorAll(
      'span[aria-hidden="true"]'
    )
    // When hideCheckbox is true, the checkbox indicator span should not be rendered
    // But the sr-only input still has aria-hidden, so we check specifically for the indicator span
    // which has specific styling classes
    const styledCheckboxes = Array.from(checkboxIndicators).filter(el =>
      el.classList.contains('h-6')
    )
    expect(styledCheckboxes.length).toBe(0)
  })

  it('renders disabled state correctly', () => {
    render(<Disabled />)
    const disabledUnchecked = screen.getByLabelText('Select disabled-unchecked')
    const disabledChecked = screen.getByLabelText('Select disabled-checked')

    expect(disabledUnchecked).toHaveAttribute('aria-checked', 'false')
    expect(disabledUnchecked).toHaveClass('opacity-50')
    expect(disabledUnchecked).toHaveAttribute('tabindex', '-1')

    expect(disabledChecked).toHaveAttribute('aria-checked', 'true')
    expect(disabledChecked).toHaveClass('opacity-50')
    expect(disabledChecked).toHaveAttribute('tabindex', '-1')
  })

  it('renders children as function correctly', () => {
    const { container } = render(<DynamicContent />)

    // Should render the dynamic content from function-as-children
    expect(container.textContent).toContain('Free Plan')
    expect(container.textContent).toContain('Premium Plan')
    expect(container.textContent).toContain('Enterprise Plan')
  })

  it('has proper accessibility attributes', () => {
    const { getByLabelText } = render(<Basic />)

    const option = getByLabelText('Select apple') as HTMLLabelElement
    expect(option).toHaveAttribute('role', 'checkbox')
    expect(option).toHaveAttribute('aria-checked')
    expect(option).toHaveAttribute('aria-label', 'Select apple')
  })
})
