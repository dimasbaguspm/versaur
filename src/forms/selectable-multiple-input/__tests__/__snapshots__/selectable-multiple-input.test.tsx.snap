// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SelectableMultipleInput > renders basic options and allows multiple selections 1`] = `
<DocumentFragment>
  <div
    class="flex flex-col"
  >
    <label
      aria-checked="false"
      aria-label="Select apple"
      class="flex gap-3 cursor-pointer border-b border-border select-none p-4 transition-colors hover:bg-neutral-light items-center"
      for="«r0»"
      role="checkbox"
      tabindex="0"
    >
      <input
        aria-checked="false"
        aria-hidden="true"
        class="sr-only"
        id="«r0»"
        type="checkbox"
        value="apple"
      />
      <span
        aria-hidden="true"
        class="relative h-6 w-6 rounded border cursor-pointer transition-all duration-200 focus:outline-none focus:ring-2 focus:ring-offset-2 appearance-none checked:bg-primary checked:border-primary focus:ring-primary/20 border-primary/40 bg-white active:scale-95 after:content-[""] after:absolute after:left-1/2 after:top-[45%] after:-translate-x-1/2 after:-translate-y-1/2 after:border-b-2 after:border-r-2 after:border-white after:rotate-45 after:opacity-0 after:transition-opacity after:duration-200 after:w-[6px] after:h-[12px]"
      />
      <span
        class="flex-1 min-w-0 text-base text-foreground"
      >
        Apple
      </span>
    </label>
    <label
      aria-checked="false"
      aria-label="Select banana"
      class="flex gap-3 cursor-pointer border-b border-border select-none p-4 transition-colors hover:bg-neutral-light items-center"
      for="«r1»"
      role="checkbox"
      tabindex="0"
    >
      <input
        aria-checked="false"
        aria-hidden="true"
        class="sr-only"
        id="«r1»"
        type="checkbox"
        value="banana"
      />
      <span
        aria-hidden="true"
        class="relative h-6 w-6 rounded border cursor-pointer transition-all duration-200 focus:outline-none focus:ring-2 focus:ring-offset-2 appearance-none checked:bg-primary checked:border-primary focus:ring-primary/20 border-primary/40 bg-white active:scale-95 after:content-[""] after:absolute after:left-1/2 after:top-[45%] after:-translate-x-1/2 after:-translate-y-1/2 after:border-b-2 after:border-r-2 after:border-white after:rotate-45 after:opacity-0 after:transition-opacity after:duration-200 after:w-[6px] after:h-[12px]"
      />
      <span
        class="flex-1 min-w-0 text-base text-foreground"
      >
        Banana
      </span>
    </label>
    <label
      aria-checked="false"
      aria-label="Select orange"
      class="flex gap-3 cursor-pointer border-b border-border select-none p-4 transition-colors hover:bg-neutral-light items-center"
      for="«r2»"
      role="checkbox"
      tabindex="0"
    >
      <input
        aria-checked="false"
        aria-hidden="true"
        class="sr-only"
        id="«r2»"
        type="checkbox"
        value="orange"
      />
      <span
        aria-hidden="true"
        class="relative h-6 w-6 rounded border cursor-pointer transition-all duration-200 focus:outline-none focus:ring-2 focus:ring-offset-2 appearance-none checked:bg-primary checked:border-primary focus:ring-primary/20 border-primary/40 bg-white active:scale-95 after:content-[""] after:absolute after:left-1/2 after:top-[45%] after:-translate-x-1/2 after:-translate-y-1/2 after:border-b-2 after:border-r-2 after:border-white after:rotate-45 after:opacity-0 after:transition-opacity after:duration-200 after:w-[6px] after:h-[12px]"
      />
      <span
        class="flex-1 min-w-0 text-base text-foreground"
      >
        Orange
      </span>
    </label>
  </div>
</DocumentFragment>
`;
